src/MagnumPlugins/KtxImporter/KtxImporter.cpp

```diff
@@ -0,0 +1,553 @@
+
+    /* Check magic string to verify this is a KTX file */
+    if(std::memcmp(&header.identifier, Implementation::KtxFileIdentifier, sizeof(header.identifier)) != 0) {
+        Error() << "Trade::KtxImporter::openData(): wrong file signature";
+        return;
+    }
+
+    Containers::Pointer<File> f{new File};
+    f->in = Containers::Array<char>{NoInit, data.size()};
+    Utility::copy(data, f->in);
+
+    /** @todo Support Basis compression */
+    const PixelFormat format = pixelFormat(Vk::PixelFormat(header.vkFormat));
+    if(format == PixelFormat{}) {
+        const CompressedPixelFormat compressedFormat = compressedPixelFormat(Vk::PixelFormat(header.vkFormat));
+        if(compressedFormat == CompressedPixelFormat{}) {
+            Error{} << "Trade::KtxImporter::openData(): unsupported format" << header.vkFormat;
+            return;
+        }
+        f->pixelFormat.compressed = compressedFormat;
+        f->compressed = true;
+    } else {
+        f->pixelFormat.uncompressed = format;
+        f->compressed = false;
+    }
+
+    /** @todo Determine if format needs swizzling */
+    f->needsSwizzle = false;
+
+    /* typeSize is the size of the format's underlying type, not the texel
+       size, e.g. 2 for RG16F. For any sane format it should be a
+       power-of-two between 1 and 8. */
+    if(header.typeSize < 1 || header.typeSize > 8 ||
+        (header.typeSize & (header.typeSize - 1)))
+    {
+        Error{} << "Trade::KtxImporter::openData(): unsupported type size" << header.typeSize;
+        return;
+    }
+    /* Block-compressed formats must have typeSize set to 1 */
+    CORRADE_INTERNAL_ASSERT(!f->compressed || header.typeSize == 1);
+    f->typeSize = header.typeSize;
+
+    /** @todo Support supercompression */
+    if(header.supercompressionScheme != 0) {
+        Error{} << "Trade::KtxImporter::openData(): supercompression is not supported";
+        return;
+    }
+
+    const Vector3ui originalSize{header.pixelWidth, header.pixelHeight, header.pixelDepth};
+    CORRADE_INTERNAL_ASSERT(originalSize.x() > 0);
+
+    if(originalSize.z() > 0) {
+        CORRADE_INTERNAL_ASSERT(originalSize.y() > 0);
+        f->dimensions = 3;
+    } else if(originalSize.y() > 0)
+        f->dimensions = 2;
+    else
+        f->dimensions = 1;
+
+    /** @todo Assert that 3D images can't have depth format */
+
+    /* Make size in each dimension at least 1 so we don't choke on size
+       calculations using product(). */
+    const Vector3i size = Math::max(Vector3i{originalSize}, Vector3i{1});
+
+    /* Number of array layers, imported as separate images */
+    const UnsignedInt numLayers = Math::max(header.layerCount, 1u);
+
+    /* Cubemap faces, either 1 or 6. KTX allows incomplete cubemaps but those
+       have to be array layers with extra metadata inside key/value data. */
+    const UnsignedInt numFaces = header.faceCount;
+    if(numFaces != 1) {
+        CORRADE_INTERNAL_ASSERT(numFaces == 6);
```

Comment:
These are all bugs in exporters, is it enough to assert here? Or should all these asserts be `Error` output? I went with `Error` for unsupported but valid scenarios or missing data (file too short etc.).

Reply:
The importer should be immune to malformed files (if you download a file and half of it is garbage, for example), so asserts should be only self-checking the code itself, not the data.

I think it could be relatively easy to test all of them using an instanced test where you have some valid file as a starting point but parametrize the file length, expected failure message and do things like "change byte at position X to value Y". Or if even that would become too involved, we can look into other ways.

Reply:
Testing the errors should be fairly straight-forward with instanced tests, I'm mostly worried about messy code because of all the checks. I might just stick most of them into a `validateHeader` function so validation and import code are less intertwined.